vmware-event-router/internal/provider/vcenter/checkpoint.go

```diff
@@ -0,0 +1,169 @@
+	// replaying the event history
+	LastEventKeyTimestamp time.Time `json:"lastEventKeyTimestamp"`
+	// timestamp (UTC) when this checkpoint was created
+	CreatedTimestamp time.Time `json:"createdTimestamp"`
+}
+
+// getCheckpoint returns a checkpoint and its full path for the given host and
+// directory. If no existing checkpoint is found, an empty checkpoint and
+// associated file is created. Thus the checkpoint might be in initialized state
+// (i.e. default values) and it is the callers responsibility to check for
+// validity using time.IsZero() on any timestamp.
+func getCheckpoint(ctx context.Context, host, dir string) (*checkpoint, string, error) {
+	var (
+		cp   *checkpoint
+		skip bool
+	)
+
+	file := fileName(host)
+	path := fullPath(file, dir)
+
+	f, err := os.Open(path)
+	defer func() {
+		_ = f.Close()
+	}()
+
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			skip = true
+			cp, err = initCheckpoint(ctx, path)
+			if err != nil {
+				return nil, "", errors.Wrap(err, "could not initialize checkpoint")
+			}
+		} else {
+			return nil, "", errors.Wrap(err, "could not configure checkpointing")
+		}
+	}
+
+	if !skip {
+		cp, err = lastCheckpoint(ctx, f)
+		if err != nil {
+			return nil, "", errors.Wrap(err, "could not retrieve last checkpoint")
+		}
+	}
+
+	return cp, path, nil
+}
+
+// initCheckpoint creates an empty checkpoint file with default values at the
+// given full file path and returns the checkpoint.
+func initCheckpoint(_ context.Context, fullPath string) (*checkpoint, error) {
+	dir := filepath.Dir(fullPath)
+
+	// check if directory already exists
+	err := os.MkdirAll(dir, 0755)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create checkpoint directory")
+	}
+
+	// create empty checkpoint
+	var cp checkpoint
+	jsonBytes, err := json.Marshal(cp)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not marshal checkpoint to JSON object")
+	}
+
+	err = ioutil.WriteFile(fullPath, jsonBytes, 0600)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not write checkpoint file")
+	}
+	return &cp, nil
+}
+
+// lastCheckpoint returns the last checkpoint for the given file
+func lastCheckpoint(_ context.Context, file io.Reader) (*checkpoint, error) {
+	var cp checkpoint
+	jsonBytes, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not read checkpoint file")
+	}
+
+	if err = json.Unmarshal(jsonBytes, &cp); err != nil {
+		return nil, errors.Wrap(err, "could not validate last checkpoint")
+	}
+
+	return &cp, nil
+}
+
+// createCheckpoint creates a checkpoint using the given file, vcenter host name
+// and checkpoint timestamp returning the created checkpoint. If lastEvent is
+// nil an errEventInvalid will be returned.
+func createCheckpoint(_ context.Context, file io.Writer, vcHost string, lastEvent types.BaseEvent, timestamp time.Time) (*checkpoint, error) {
+	// won't catch the case when the value is not pointer
```

Comment:
> It won't catch non-pointers, but won't it also cause a panic if a struct is passed by value? I'm not very good with reflect, so I may be wrong about this.

This is intentional since I do want the code to panic in case of `lastEvent` being non-pointer (which should never be the case, though). If it is, e.g. govmomi changes, we'll catch this.

> I wonder if doing a type switch might work? I guess it would only work if you know what types will implement the types.BaseEvent interface. I think only types.Event implements types.BaseEvent. Is that right?

The link you gave actually names a few problems of the alternatives I have researched and thus decided to panic to clearly catch a deviation of the assumptions we're making against govmomi.

> But I think using reflect in production code is a bad idea, especially because of high potential to panic.

Actually, I disagree based on the rational shared earlier. If our assumptions break about an interface implementation, we should panic instead of letting an API change slip through which is super hard to spot. Also, I don't agree with the general statement `using reflect in production code is a bad idea`. That would exclude `json` from the standard library ;)

Reply:
I see your reasoning there as to why panicking is ok in this case.

As for `using reflect in production code is a bad idea`, I agree it's an incorrect statement for the example you gave. Also, it seems reflection is often used in frameworks. 

I've read that reflection should only be used in application code as a last resort, I guess due to panics. But since you prefer the code to panic in this case, it seems ok to use the reflect package.

Reply:
In my opinion, in general if its there it should be used, but this is subjective. I've been advised to avoid usage on reflect only to avoid adding excess libraries when I've wanted to compare structs etc. because you can pretty much do that with built in go functions.

On the panic or not, if we always expect the value to be a pointer and it isn't this isn't something the user can do and it is [unexpected behavior](https://gobyexample.com/panic) and is worthy of panic. Otherwise I think that it would be worth elaborating on why we won't catch the case e.g.:
```
// won't catch the case when the value is not pointer
// since non-pointer value would be unexpected 
// behaivior from govmomi
```
Because it can give a context in case a future developer gets the panic while updating govmomi.